### PR TITLE
fix(types): add missing model hooks types

### LINKED ISSUE #14955: validationFailed hook is missing in type definitions
## Issue Creation Checklist

- [x] I have read the [contribution guidelines](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)

## Issue Description

The TS type definition for the `validationFailed` hook is missing.

### Additional context

The signature should be added to `ModelHooks` interface (allowing to add it with `addHook()` as  well). There might be other places but this is the one I could find.

### Is this issue dialect-specific?

- [x] No. This issue is relevant to Sequelize as a whole.
- [ ] Yes. This issue only applies to the following dialect(s): XXX, YYY, ZZZ
- [ ] I don't know.

### Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [ ] Yes, I have the time but I don't know how to start, I would need guidance.
- [x] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

### PR DESCRIPTION
**Pull request by fzn0x**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [ ] Have you added new tests to prevent regressions?
- [ ] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Closes #14955

### Todos

- [x] Add missing `validationFailed`,`beforeAssociate` and `afterAssociate` types.
- [ ] Maybe add missing types tests?

### CODE REVIEW

**Comment by sdepold** [comment]
Should this be covered by tests?

**Comment by fzn0x** [comment]
> Should this be covered by tests?

Since I found the tests for hooks types https://github.com/sequelize/sequelize/blob/missing-model-hooks-types/test/types/hooks.ts I think it would be a yes, what do you think guys? @ephys @WikiRik , also ephys already back too so I can requested review from her whether is it should or not.

**Comment by WikiRik** [comment]
Yes, some typing tests would be nice

**Comment by fzn0x** [comment]
Let's see if it's "correctly" passed

**Comment by WikiRik** [comment]
https://github.com/sequelize/sequelize/pull/14959 is related to this as well I think. @ephys looked into expect-type as well for the previous version

**Comment by fzn0x** [comment]
Ouch, seems checks hangs

**Comment by fzn0x** [comment]
Yay it's passed!

**Review by ephys — APPROVED** [review]

File: src/hooks.d.ts [hunk 1]
```diff
@@ -62,6 +68,8 @@ export interface ModelHooks<M extends Model = Model, TAttributes = any> {
   afterBulkSync(options: SyncOptions): HookReturn;
   beforeQuery(options: QueryOptions, query: AbstractQuery): HookReturn;
   afterQuery(options: QueryOptions, query: AbstractQuery): HookReturn;
+  beforeAssociate(data: BeforeAssociateEventData, options: AssociationOptions<any>): HookReturn;
```
Comment (ephys) [inline src/hooks.d.ts:71]: I'd prefer to go with the safer "unknown"

```suggestion
  beforeAssociate(data: BeforeAssociateEventData, options: AssociationOptions<unknown>): HookReturn;
```

File: src/hooks.d.ts [hunk 2]
```diff
@@ -62,6 +68,8 @@ export interface ModelHooks<M extends Model = Model, TAttributes = any> {
...
+  afterAssociate(data: AfterAssociateEventData, options: AssociationOptions<any>): HookReturn;
```
Comment (ephys) [inline src/hooks.d.ts:72]:
```suggestion
  afterAssociate(data: AfterAssociateEventData, options: AssociationOptions<unknown>): HookReturn;
```

File: test/types/hooks.ts [hunk 3]
```diff
@@ -1,14 +1,25 @@
...
 import type { AbstractQuery } from '@sequelize/core/_non-semver-use-at-your-own-risk_/dialects/abstract/query.js';
 import type { ModelHooks } from '@sequelize/core/_non-semver-use-at-your-own-risk_/hooks.js';
+import type { ValidationOptions } from '@sequelize/core/_non-semver-use-at-your-own-risk_/instance-validator';
 import { expectTypeOf } from 'expect-type';
 import type { SemiDeepWritable } from './type-helpers/deep-writable';
 
 {
   class TestModel extends Model {}
 
   const hooks: Partial<ModelHooks> = {
```
Comment (ephys) [inline test/types/hooks.ts:17]:
```suggestion
  const hooks: Partial<ModelHooks<TestModel>> = {
```

**Review by ephys — CHANGES_REQUESTED** [review]
Tiny details

File: src/hooks.d.ts (hunk 1 above)
Comment (fzn0x) [reply src/hooks.d.ts:71]: Makes sense, I figured out `unknown` is better for this case than `any`

File: src/hooks.d.ts (hunk 2 above)
Comment (fzn0x) [reply src/hooks.d.ts:72]: Thanks

File: test/types/hooks.ts (hunk 3 above)
Comment (fzn0x) [reply test/types/hooks.ts:17]: Thank you

**Comment by fzn0x** [comment]
I'm noticed that we still have <any> for AssociationOptions here: https://github.com/sequelize/sequelize/blob/8ce01106a34d26dc0e3bec605e9fb7e0ac6c6895/src/model.d.ts#L3230

Is it okay to include the changes here?

**Comment by fzn0x** [comment]
I'm noticed that we still have <any> for AssociationOptions here: https://github.com/sequelize/sequelize/blob/8ce01106a34d26dc0e3bec605e9fb7e0ac6c6895/src/model.d.ts#L3230

Is it okay to include the changes here (since I think it's still related to `missing types` context)? @ephys

**Comment by fzn0x** [comment]
I'm noticed that we still have `<any>` for AssociationOptions here: https://github.com/sequelize/sequelize/blob/8ce01106a34d26dc0e3bec605e9fb7e0ac6c6895/src/model.d.ts#L3230

Is it okay to include the changes here (since I think it's still related to `missing types` context)? @ephys

**Comment by fzn0x** [comment]
Unfortunately it didn't passed with `unknown` data type

**Comment by ephys** [comment]
I think that could be changed to unknown too

**Review by ephys — APPROVED** [review]

**Pull request merged**